### ISSUE TITLE
Implement Stream Out / Transform Feedback shaders

### ISSUE THREAD

**Issue by achurch** [decaf-emu/decaf-emu#234]
Xenoblade gives this assertion failure for the shader pair below. The vertex shader maps semantic IDs 64-66 to outputs 0-2, but the pixel shader looks up semantic ID 0, triggering the assertion.

I note that the vertex shader has a POS0 output in addition to PARAM0-PARAM2; could this be what the pixel shader is trying to read from? (Though on second thought, that doesn't make sense because the POS0 output is a constant {0,0,0,1}...)

Vertex shader: (this looks like it's actually the same shader as in issue #221)
```
00 CALL_FS NO_BARRIER

01 ALU ADDR(32) CNT(25) KCACHE0(cb2, 0 to 15)
    0 x: DOT4            ____, R2.x, 1.0f
      y: DOT4            ____, R2.y, 1.0f
      z: DOT4            ____, R2.z, 1.0f
      w: DOT4            ____, (0x80000000, -0), 0.0f
      t: MULLO_INT       R3.y, R1.x, (0x00000003, 4.2039e-45) SCL_210

    1 x: ADD_INT         R0.x, PS0, (0x00000002, 2.8026e-45)
      y: ADD             R15.y, -PV0.x, 1.0f CLAMP
      z: MUL             R15.z, KCACHE0[7].w, 1.0f
      w: ADD_INT         R2.w, PS0, 1
      t: MULLO_INT       R0.z, R1.y, (0x00000003, 4.2039e-45) SCL_210

    2 x: MUL             R15.x, KCACHE0[8].w, 1.0f
      y: ADD_INT         R0.y, PS1, 1
      z: MUL             R16.z, KCACHE0[9].w, 1.0f
      w: ADD_INT         R0.w, PS1, (0x00000002, 2.8026e-45)
      t: MULLO_INT       R3.x, R1.z, (0x00000003, 4.2039e-45) SCL_210

    3 x: MOV             R1.x, 0.0f
      y: ADD_INT         R1.y, PS2, (0x00000002, 2.8026e-45)
      z: ADD_INT         R1.z, PS2, 1
      w: MOV             R1.w, (0x3F800000, 1)
      t: MULLO_INT       R3.z, R1.w, (0x00000003, 4.2039e-45) SCL_210

02 TEX ADDR(192) CNT(8)

    4    FETCH           R4.xyzw, R0.x, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

    5    FETCH           R5.xyzw, R0.w, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

    6    FETCH           R6.xyzw, R2.w, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

    7    FETCH           R7.xyzw, R0.y, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

    8    FETCH           R8.xyzw, R1.y, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

    9    FETCH           R9.xyzw, R3.y, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

   10    FETCH           R10.xyzw, R0.z, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

   11    FETCH           R11.xyzw, R1.z, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

03 ALU ADDR(57) CNT(11)
   12 x: MUL             R0.x, R2.x, R4.y
      y: ADD_INT         R3.y, R3.z, 1
      z: MUL             R1.z, R2.x, R4.x
      w: ADD_INT         R2.w, R3.z, (0x00000002, 2.8026e-45)
      t: MUL             R0.w, R2.x, R4.z

   13 x: MUL             R4.x, R2.x, R4.w VEC_021
      y: MUL             R4.y, R2.y, R5.x
      z: MUL             R4.z, R2.y, R5.y
      w: MUL             R4.w, R2.y, R5.z
      t: MUL             R5.w, R2.y, R5.w

04 TEX ADDR(208) CNT(4)

   14    FETCH           R12.xyzw, R2.w, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

   15    FETCH           R13.xyzw, R3.x, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

   16    FETCH           R14.xyzw, R3.y, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

   17    FETCH           R3.xyzw, R3.z, b129 NO_INDEX_OFFSET FMT_FROM_FETCH_CONSTANT MEGA(16) OFFSET(0)

05 EXP_DONE POS0, R1.xxxw

06 ALU ADDR(68) CNT(112) KCACHE0(cb2, 0 to 15)
   18 x: ADD             R125.x, R1.z, R4.y VEC_021
      y: ADD             R126.y, R0.w, R4.w
      z: ADD             R125.z, R0.x, R4.z
      w: ADD             R125.w, R4.x, R5.w VEC_120
      t: MUL             R126.z, R2.x, R6.y

   19 x: MUL             R127.x, R2.x, R6.x
      y: MUL             R127.y, R2.x, R6.z
      z: MUL             R127.z, R2.y, R7.x VEC_021
      w: MUL             R127.w, R2.x, R6.w
      t: MUL             R126.x, R2.y, R7.y

   20 x: MUL             R124.x, R2.z, R8.y VEC_201
      y: MUL             ____, R2.y, R7.w VEC_120
      z: MUL             ____, R2.y, R7.z VEC_102
      w: MUL             R126.w, R2.z, R8.x VEC_201
      t: MUL             R124.z, R2.z, R8.z

   21 x: ADD             R126.x, R127.x, R127.z
      y: MUL             ____, R2.z, R8.w
      z: ADD             R127.z, R126.z, R126.x VEC_210
      w: ADD             R124.w, R127.y, PV20.z
      t: ADD             R8.w, R127.w, PV20.y

   22 x: ADD             R8.x, R125.z, R124.x
      y: ADD             R124.y, R125.w, PV21.y
      z: ADD             R8.z, R126.y, R124.z
      w: ADD             R7.w, R125.x, R126.w
      t: MUL             R124.x, R2.x, R9.y

   23 x: MUL             R125.x, R2.y, R10.x
      y: MUL             R126.y, R2.x, R9.w
      z: MUL             R125.z, R2.x, R9.z
      w: MUL             R126.w, R2.x, R9.x VEC_021
      t: MUL             R125.w, R2.y, R10.y

   24 x: MUL             R127.x, R2.y, R10.z VEC_102
      y: MUL             R127.y, R2.z, R11.x VEC_201
      z: MUL             R124.z, R2.y, R10.w VEC_120
      w: MUL             R127.w, R2.z, R11.y VEC_201
      t: MUL             R10.x, R2.z, R11.z

   25 x: MUL             R11.x, R15.y, R12.x
      y: MUL             R125.y, R15.y, R12.y
      z: MUL             R126.z, R2.z, R11.w VEC_021
      w: MUL             R11.w, R15.y, R12.z
      t: MUL             R12.w, R15.y, R12.w

   26 x: ADD             R127.x, R126.w, R125.x
      y: ADD             R126.y, R125.z, R127.x VEC_021
      z: ADD             R127.z, R124.x, R125.w VEC_021
      w: ADD             R127.w, R126.y, R124.z
      t: ADD             R125.x, R127.z, R127.w

   27 x: ADD             R124.x, R7.w, R11.x
      y: ADD             R11.y, R8.w, R126.z VEC_102
      z: ADD             R126.z, R124.w, R10.x VEC_201
      w: ADD             R126.w, R126.x, R127.y VEC_201
      t: MUL             R127.y, R2.z, R13.y

   28 x: MUL             R126.x, R2.z, R13.x VEC_201
      y: ADD             R12.y, R8.z, R11.w
      z: ADD             R124.z, R8.x, R125.y VEC_120
      w: ADD             R125.w, R124.y, R12.w VEC_021
      t: MUL             R124.w, R2.z, R13.z

   29 x: MUL             ____, R2.z, R13.w VEC_021
      y: MUL             R125.y, R15.y, R14.y
      z: MUL             R125.z, R15.y, R14.z
      w: MUL             ____, R15.y, R14.x
      t: MUL             R124.y, R15.y, R14.w

   30 x: ADD             R127.x, R127.z, R127.y
      y: ADD             R127.y, R127.w, PV29.x
      z: ADD             R127.z, R126.y, R124.w
      w: ADD             R127.w, R127.x, R126.x
      t: ADD             R13.x, R126.w, PV29.w

   31 x: MUL             R2.x, R124.x, KCACHE0[7].z
      y: ADD             R11.y, R126.z, R125.z
      z: ADD             R2.z, R125.x, R125.y VEC_120
      w: ADD             R13.w, R11.y, R124.y
      t: MUL             R126.x, R124.z, KCACHE0[7].z

   32 x: MUL             R125.x, R15.y, R3.z
      y: MUL             R125.y, R15.y, R3.x
      z: MUL             R126.z, R15.y, R3.w
      w: MUL             R126.w, R15.y, R3.y
      t: MUL             R124.w, R12.y, KCACHE0[7].z

   33 x: MUL             R15.x, R124.x, KCACHE0[8].z
      y: MUL             R124.y, R12.y, KCACHE0[8].z
      z: MULADD          R125.z, KCACHE0[7].z, R125.w, R15.z
      w: MUL             R3.w, R124.z, KCACHE0[8].z
      t: MULADD          R126.y, KCACHE0[8].z, R125.w, R15.x VEC_021

   34 x: MULADD          R124.x, KCACHE0[9].z, R125.w, R16.z
      y: MUL             R125.y, R12.y, KCACHE0[9].z
      z: MUL             R124.z, R124.z, KCACHE0[9].z
      w: MUL             R127.w, R124.x, KCACHE0[9].z
      t: ADD             R3.x, R127.w, R125.y

   35 x: MULADD          R127.x, KCACHE0[7].y, R2.z, R126.x
      y: ADD             R127.y, R127.z, R125.x
      z: ADD             R126.z, R127.x, R126.w
      w: ADD             R125.w, R127.y, R126.z VEC_021
      t: MULADD          R126.w, KCACHE0[7].y, R11.y, R124.w VEC_021

   36 x: MULADD          R126.x, KCACHE0[8].y, R13.x, R15.x
      y: MULADD          R124.y, KCACHE0[7].y, R13.x, R2.x VEC_210
      z: MULADD          R125.z, KCACHE0[7].y, R13.w, R125.z
      w: MULADD          R124.w, KCACHE0[8].y, R2.z, R3.w
      t: MULADD          R127.z, KCACHE0[8].y, R11.y, R124.y VEC_120

   37 x: MULADD          R124.x, KCACHE0[9].y, R11.y, R125.y
      y: MULADD          R126.y, KCACHE0[8].y, R13.w, R126.y VEC_120
      z: MULADD          R124.z, KCACHE0[9].y, R2.z, R124.z
      w: MULADD          R127.w, KCACHE0[9].y, R13.x, R127.w VEC_201
      t: MULADD          R125.x, KCACHE0[9].y, R13.w, R124.x VEC_021

   38 x: MULADD          R2.x, KCACHE0[7].x, R3.x, R124.y
      y: MULADD          R2.y, KCACHE0[7].x, R126.z, R127.x VEC_120
      z: MULADD          R2.z, KCACHE0[7].x, R127.y, R126.w
      w: MULADD          R2.w, KCACHE0[7].x, R125.w, R125.z VEC_201
      t: MULADD          R11.x, KCACHE0[8].x, R3.x, R126.x VEC_120

   39 x: MULADD          R3.x, KCACHE0[9].x, R3.x, R127.w
      y: MULADD          R11.y, KCACHE0[8].x, R126.z, R124.w VEC_210
      z: MULADD          R11.z, KCACHE0[8].x, R127.y, R127.z VEC_120
      w: MULADD          R11.w, KCACHE0[8].x, R125.w, R126.y VEC_210
      t: MULADD          R3.y, KCACHE0[9].x, R126.z, R124.z VEC_120

   40 z: MULADD          R3.z, KCACHE0[9].x, R127.y, R124.x
      w: MULADD          R3.w, KCACHE0[9].x, R125.w, R125.x VEC_021

07 EXP PARAM0, R2.xyzw

08 MEM_STREAM0 MEM_UNKNOWN_FORMAT

09 EXP PARAM1, R11.xyzw NO_BARRIER

10 MEM_STREAM0 MEM_UNKNOWN_FORMAT

11 EXP_DONE PARAM2, R3.xyzw NO_BARRIER

12 MEM_STREAM0 MEM_UNKNOWN_FORMAT

13 ALU ADDR(180) CNT(1)
   41 x: NOP             ____

14 NOP
```

Pixel shader:
```
00 TEX ADDR(48) CNT(1) VALID_PIX

    0    SAMPLE          R0.___x, R0.xy0x, t0, s0

01 ALU ADDR(32) CNT(1)
    1 x: MOV             R0.x, 0.0f

02 EXP_DONE PIXEL0, R0.xxxw
```

**Comment by achurch** [comment]
Looking more closely, the vertex shader ends with some MEM_STREAM0 instructions (I've currently turned those into no-ops for testing so the GLSL translator doesn't assert), and the game spams calls to unimplemented GX2*StreamOut* right before this assertion is hit. Is it possible this is a transform feedback operation and the pixel shader isn't in fact used at all?

**Comment by exjam** [comment]
Yeah I have not actually looked into anything stream out related, but it is equivalent to OpenGL transform feedback from what I can tell, which would indeed mean the pixel shader is never actually used. Only a vertex shader is used with the output stored into a buffer. Will require some work to figure out how stream out works in gx2 to hook this up though.

**Comment by achurch** [comment]
FYI, I've done some work on this: https://github.com/achurch/decaf-emu/tree/streamout
Not done yet (getting some OpenGL errors, and I also want to look at GX2QueryBegin/End), but at least there's a lot less "unimplemented" spam now.

**Comment by exjam** [comment]
Looks like a great start!

1. A bit dubious about the modification of GX2SetVertexShader and GX2SetGeometryShader, we reversed that straight out of gx2, so you should reconsider why you removed them.

2. A quick look at gx2.rpl shows GX2SetStreamOutContext is using a pm4 packet with C0043400 which is a  STRMOUT_BUFFER_UPDATE, we should always be using the real pm4 commands where possible. Decaf prefixed pm4 commands are for things which do not use the standard pm4 packets on hardware, so we kind of cheat and inject our custom commands into the pm4 stream to get the message through to the graphics driver.

I can take a look at reversing all the correct pm4 packets for the stream out methods later tonight if you want, I have done a lot so its not too hard to me at this point.

**Comment by achurch** [comment]
>A bit dubious about the modification of GX2SetVertexShader and GX2SetGeometryShader

It made no sense to me to set just the stride and not the buffer and size, but on second thought, I guess that matches how in GLSL you can specify the buffer stride as part of the shader, so I'll revert that.

>A quick look at gx2.rpl shows GX2SetStreamOutContext is using a pm4 packet with C0043400 which is a STRMOUT_BUFFER_UPDATE

I have no idea what gx2.rpl does (I wrote those based on the game's call pattern), so any info you have on that front is appreciated. (: That said, based on the Mesa R600 code (about which I have no idea of the quality, granted), STRMOUT_BUFFER_UPDATE doesn't include a field for primitive mode, so we'd need a custom packet anyway.

It's also possible the third parameter to GX2SetStreamOutContext isn't a primitive mode - Xenoblade passes 1 which is POINTLIST, and it only ever draws points when doing transform feedback, but it could just as easily be a flag of some sort. In that case I guess we'd need some other tricks to pass the correct mode to glBeginTransformFeedback.

**Comment by exjam** [comment]
Here is the correct reversed GX2 -> pm4, you correctly guessed GX2SetStreamOutEnable at least :+1: others needed a few modifications. The 3rd param for GX2SetStreamOutContext is interesting, I believe its to indicate the source offset of the stream out buffer based off what the AMD docs say and what the code does, it's funny if value == 2 or FromOffset then it reuses the 2nd arg (pointer to stream) to actually be the offset instead, pretty hacky.

https://github.com/decaf-emu/decaf-emu/commit/cca50bfef39a54456a7b5bc1838f494d498d0b10

**Comment by achurch** [comment]
Nice, thanks! I figured out the PM4 part (was going to push it to github but got sidetracked with some other stuff...) but nice find on the 3rd arg to GX2SetStreamOutContext.

**Comment by achurch** [comment]
Hmm, are you sure about the packet format? The Mesa code suggests it only has 4 data words after the control word:
```
radeon_emit(cs, PKT3(PKT3_STRMOUT_BUFFER_UPDATE, 4, 0));
radeon_emit(cs, STRMOUT_SELECT_BUFFER(i) |
            STRMOUT_OFFSET_SOURCE(STRMOUT_OFFSET_FROM_PACKET)); /* control */
radeon_emit(cs, 0); /* unused */
radeon_emit(cs, 0); /* unused */
radeon_emit(cs, t[i]->b.buffer_offset >> 2); /* buffer offset in DW */
radeon_emit(cs, 0); /* unused */
```

**Comment by exjam** [comment]
Ah yeah, buffer offset & srcAddrLo is the same field... I can confirm this by looking at gx2.rpl.

This page break confused me from the AMD docs! https://puu.sh/qpzlP/2b05d3d8a5.png

**Comment by achurch** [comment]
Hah, formatting fail! Okay, I think I've got it all covered -- going to take a look at GX2QueryBegin/End since that seems to be the last source of log spam.

**Comment by exjam** [comment]
Well it's not trivial, a quick look shows it is like:

loop { MEM_WRITE, PFP_SYNC_ME },
SET_CONTEXT_REG DB_RENDER_CONTROL,
EVENT_WRITE

**Comment by achurch** [comment]
I'll take a stab at it from the client side, though Xenoblade's query processing routine is like 3 pages of rlwinm/rlwimi instructions so it may take a while to deobfuscate it, lol. In the meantime I'll compact the transform feedback changes and put up a PR for you to look over.

**Comment by exjam** [comment]
Here is GX2QueryBegin / GX2QueryEnd, it's a bit weird... Not sure how to implement it from the OpenGL side, maybe you could find more information in Mesa.

https://github.com/decaf-emu/decaf-emu/tree/query
https://github.com/decaf-emu/decaf-emu/commit/532eaaa52766d8bfd861bff7c2e0c955985c8073

**Comment by achurch** [comment]
Fixed by #242.
